src/lib.rs

```diff
@@ -0,0 +1,168 @@
+    }
+
+    /// Try to find quality of the proof at `challenge_index` if proof exists
+    pub fn find_quality(&self, challenge_index: u32) -> Option<Quality<'_>> {
+        let mut bytes = [0u8; 32];
+        // SAFETY: Called with prover that is still alive
+        unsafe {
+            ffi::subspace_chiapos_find_quality(self.prover, challenge_index, bytes.as_mut_ptr())
+        }
+        .then_some(Quality {
+            bytes,
+            challenge_index,
+            table: self,
+        })
+    }
+}
+
+/// Check whether proof created earlier is valid
+pub fn is_proof_valid(seed: &[u8; 32], challenge_index: u32, proof: &[u8; K as usize * 8]) -> bool {
+    // SAFETY: Called with valid pointer to seed and proof with correct size
+    unsafe {
+        ffi::subspace_chiapos_is_proof_valid(K, seed.as_ptr(), challenge_index, proof.as_ptr())
+    }
+}
+
+mod ffi {
+    use std::ffi::c_void;
+
+    #[repr(transparent)]
+    #[derive(Debug, Copy, Clone)]
+    pub struct Table(*const c_void);
+
+    unsafe impl Send for Table {}
+    unsafe impl Sync for Table {}
+
+    #[repr(transparent)]
+    #[derive(Debug, Copy, Clone)]
+    pub struct Prover(*const c_void);
+
+    unsafe impl Send for Prover {}
+    unsafe impl Sync for Prover {}
+
+    extern "C" {
+        // Create new table for K with 32 bytes seed
+        pub(super) fn subspace_chiapos_create_table(k: u8, seed: *const u8) -> Table;
+
+        pub(super) fn subspace_chiapos_free_table(table: Table);
+
+        pub(super) fn subspace_chiapos_create_prover(table: Table) -> Prover;
+
+        pub(super) fn subspace_chiapos_free_prover(prover: Prover);
+
+        // Prover is the same as created by `create_prover` above.
+        //
+        // On success writes `32` bytes and returns `true`, returns `false` otherwise.
+        pub(super) fn subspace_chiapos_find_quality(
+            prover: Prover,
+            challenge_index: u32,
+            quality: *mut u8,
+        ) -> bool;
+
+        // Prover is the same as created by `create_prover` above.
+        //
+        // On success writes `k*8` bytes and returns `true`, returns `false` otherwise.
+        pub(super) fn subspace_chiapos_create_proof(
+            prover: Prover,
+            challenge_index: u32,
+            proof: *mut u8,
+        ) -> bool;
+
+        // Check if proof is valid for K with 32 byte seed and challenge/proof form `create_proof` above.
+        pub(super) fn subspace_chiapos_is_proof_valid(
+            k: u8,
+            seed: *const u8,
+            challenge_index: u32,
+            proof: *const u8,
+        ) -> bool;
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const SEED: [u8; 32] = [
+        35, 2, 52, 4, 51, 55, 23, 84, 91, 10, 111, 12, 13, 222, 151, 16, 228, 211, 254, 45, 92,
+        198, 204, 10, 9, 10, 11, 129, 139, 171, 15, 23,
+    ];
+
+    #[test]
+    fn basic() {
+        let table = Table::generate(&SEED);
+
+        assert!(table.find_quality(0).is_none());
```

Comment:
A minor question: does quality/proof never exist for `challenge_index = 0`, or is it a coincidence? I'd add a note to the spec to have `start_index` and s-bucket numeration start at 1 or be one-off

Reply:
Have not tried other seeds, why does it matter?

Reply:
Pedantry, mostly. If the table never has an entry at 0, it makes no sense to query for 0 for every record. According to [design doc](https://github.com/Chia-Network/chiapos/blob/main/documents/proof_of_space.md#full-algorithm) it should exist. This is not a concern for this PR just some minor thing to check.